src/DataTypes/Serializations/SerializationDynamic.cpp

```diff
@@ -364,9 +362,6 @@ void SerializationDynamic::serializeBinaryBulkWithMultipleStreamsAndCountTotalSi
     if (!variant_info.variant_type->equals(*dynamic_state->variant_type))
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Mismatch of internal columns of Dynamic. Expected: {}, Got: {}", dynamic_state->variant_type->getName(), variant_info.variant_type->getName());
 
-    if (column_dynamic.getMaxDynamicTypes() != dynamic_state->max_dynamic_types)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "Mismatch of max_dynamic_types parameter of Dynamic. Expected: {}, Got: {}", dynamic_state->max_dynamic_types, column_dynamic.getMaxDynamicTypes());
```

Comment:
Now the previous check for internal variant type equality is enough. It can happen that during `Alter` the internal value of `max_dynamic_types` is different in first serialized column (because for all next columns we use `takeDynamicStructureFromSourceColumns` that sets internal `max_dynamic_types` to the actual number of variants). And this is why I had to change the serialization a bit. Now instead of `max_dynamic_types` parameter we always serialize the actual number of variants in the structure stream (actually, only parts with 0 level could have `max_dynamic_types` different from the number of variants, because during merge we use `takeDynamicStructureFromSourceColumns` and set `max_dynamic_types` to the number of variants)